diesel/src/upsert/on_conflict_target.rs

```diff
@@ -0,0 +1,97 @@
+use backend::{Backend, SupportsOnConflictClause};
+use expression::SqlLiteral;
+use query_builder::*;
+use query_source::Column;
+use result::QueryResult;
+
+#[cfg(feature = "postgres")]
+pub use pg::on_constraint;
+
+#[doc(hidden)]
+pub trait OnConflictTarget<Table> {}
+
+#[doc(hidden)]
+#[derive(Debug, Clone, Copy)]
+pub struct NoConflictTarget;
+
+impl<DB> QueryFragment<DB> for NoConflictTarget
+where
+    DB: Backend + SupportsOnConflictClause,
+{
+    fn walk_ast(&self, _: AstPass<DB>) -> QueryResult<()> {
+        Ok(())
+    }
+}
+
+impl<Table> OnConflictTarget<Table> for NoConflictTarget {}
+
+#[doc(hidden)]
+#[derive(Debug, Clone, Copy)]
+pub struct ConflictTarget<T>(pub T);
+
+impl<DB, T> QueryFragment<DB> for ConflictTarget<T>
+where
+    DB: Backend + SupportsOnConflictClause,
+    T: Column,
+{
+    fn walk_ast(&self, mut out: AstPass<DB>) -> QueryResult<()> {
+        out.push_sql(" (");
+        try!(out.push_identifier(T::NAME));
+        out.push_sql(")");
+        Ok(())
+    }
+}
+
+impl<T> OnConflictTarget<T::Table> for ConflictTarget<T>
+where
+    T: Column,
+{
+}
+
+impl<DB, ST> QueryFragment<DB> for ConflictTarget<SqlLiteral<ST>>
+where
+    DB: Backend + SupportsOnConflictClause,
+    SqlLiteral<ST>: QueryFragment<DB>,
+{
+    fn walk_ast(&self, mut out: AstPass<DB>) -> QueryResult<()> {
+        out.push_sql(" ");
+        self.0.walk_ast(out.reborrow())?;
+        Ok(())
+    }
+}
+
+impl<Tab, ST> OnConflictTarget<Tab> for ConflictTarget<SqlLiteral<ST>> {}
+
+macro_rules! on_conflict_tuples {
+    ($($col:ident),+) => {
+        impl<DB, T, $($col),+> QueryFragment<DB> for ConflictTarget<(T, $($col),+)> where
+            DB: Backend + SupportsOnConflictClause,
+            T: Column,
+            $($col: Column<Table=T::Table>,)+
+        {
+            fn walk_ast(&self, mut out: AstPass<DB>) -> QueryResult<()> {
+                out.push_sql(" (");
+                try!(out.push_identifier(T::NAME));
+                $(
+                    out.push_sql(", ");
+                    try!(out.push_identifier($col::NAME));
+                )+
+                out.push_sql(")");
+                Ok(())
+            }
+        }
+
+        impl<T, $($col),+> OnConflictTarget<T::Table> for ConflictTarget<(T, $($col),+)> where
+            T: Column,
+            $($col: Column<Table=T::Table>,)+
+        {
+        }
+    }
+}
+
+on_conflict_tuples!(U);
```

Comment:
Unrelated (so not needed for this PR): We should replace this with a [`__diesel_for_each_tuple`](https://github.com/diesel-rs/diesel/blob/master/diesel/src/macros/tuples.rs#L3) call. 
(This is more or less as side note for the core team)